src/amp/lib/experiment.ts

```diff
@@ -0,0 +1,150 @@
+
+interface VariantConfig {
+    proportion: number;
+    style: string;
+}
+
+interface ExperimentCollection<ExperimentObject> {
+    [experimentName: string]: ExperimentObject;
+}
+
+export type AmpModelCollection = ExperimentCollection<AmpModel>;
+
+export type StyledModelCollection = ExperimentCollection<StyledExperimentModel>;
+
+type StylesCollection = ExperimentCollection<{ [variantName: string]: string }>;
+
+const extractFromVariants = (
+    variants: { [variantName: string]: VariantConfig },
+    fieldName: keyof VariantConfig,
+): { [variantName: string]: string | number } => {
+    return Object.entries(variants).reduce((acc, [variantName, config]) => {
+        acc = { ...acc, ...{ [variantName]: config[fieldName] } };
+        return acc;
+    }, {});
+};
+
+const isValidAmpModelVariant = (variant: {
+    [variantName: string]: string | number;
+}): variant is { [variantName: string]: number } => {
+    return Object.values(variant).every(x => typeof x === 'number');
+};
+
+const isValidStyleVariant = (variant: {
+    [variantName: string]: string | number;
+}): variant is { [variantName: string]: string } => {
+    return Object.values(variant).every(x => typeof x === 'string');
+};
+
+export const extractExperimentModels = (
+    fullConfig: StyledModelCollection,
+): AmpModelCollection => {
+    return Object.entries(fullConfig).reduce(
+        (extractedModels, [experimentName, styledModel]) => {
+            const { variants, ...configWithoutVariants } = styledModel;
+            const proportions = extractFromVariants(variants, 'proportion');
+            if (isValidAmpModelVariant(proportions)) {
+                extractedModels[experimentName] = {
+                    ...configWithoutVariants,
+                    ...{
+                        variants: proportions,
+                    },
+                };
+            }
+            return extractedModels;
+        },
+        {} as AmpModelCollection,
+    );
+};
+
+export const extractExperimentStyles = (
+    fullConfig: StyledModelCollection,
+): StylesCollection => {
+    return Object.entries(fullConfig).reduce(
+        (extractedModels, [experimentName, styledModel]) => {
+            const { variants } = styledModel;
+            const styles = extractFromVariants(variants, 'style');
+            if (isValidStyleVariant(styles)) {
+                extractedModels[experimentName] = { ...styles };
+            }
+            return extractedModels;
+        },
+        {} as StylesCollection,
+    );
+};
+
+export const getActiveExperiments = (
+    abTestObject: StyledModelCollection,
+    switches: { [key: string]: boolean },
+): StyledModelCollection => {
+    return Object.entries(abTestObject).reduce(
+        (acc, [testName, testConfig]) => {
+            if (switches[testName]) {
+                acc[testName] = testConfig;
+            }
+            return acc;
+        },
+        {} as StyledModelCollection,
+    );
+};
```

Comment:
```suggestion
): StyledModelCollection => Object.entries(abTestObject).reduce(
      (acc, [testName, testConfig]) => {
          if (switches[testName]) {
              return {
                ...acc,
                [testName]: testConfig,
              };
          }
          return acc;
      },
      {} as StyledModelCollection,
  );
```

Reply:
this could also be written as a ternary expression which would be shorter but not sure if it's more or less readable...
```suggestion
): StyledModelCollection => Object.entries(abTestObject).reduce(
      (acc, [testName, testConfig]) => switches[testName] ? {
                ...acc,
                [testName]: testConfig,
              } : acc,
      {} as StyledModelCollection,
  );
```

Reply:
Thanks. I took a lot of what you suggested and applied it to the recommendations from the other reviewers. Particularly glad to know that I don't have to build an inner object to spread and combine into a new one.